File: src/main/java/me/centralhardware/znatoki/telegram/statistic/telegram/CommandHandler/statisticCommand/BaseReport.java
package me.centralhardware.znatoki.telegram.statistic.telegram.CommandHandler.statisticCommand;

import me.centralhardware.znatoki.telegram.statistic.entity.postgres.Role;
import me.centralhardware.znatoki.telegram.statistic.mapper.postgres.UserMapper;
import me.centralhardware.znatoki.telegram.statistic.telegram.fsm.Storage;
import me.centralhardware.znatoki.telegram.statistic.mapper.postgres.ServiceMapper;
import me.centralhardware.znatoki.telegram.statistic.service.TelegramService;
import me.centralhardware.znatoki.telegram.statistic.telegram.CommandHandler.CommandHandler;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.methods.send.SendDocument;
import org.telegram.telegrambots.meta.api.objects.InputFile;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.User;

import java.io.File;
import java.util.List;
import java.util.function.Function;

@Component
public abstract class BaseReport extends CommandHandler {

    @Autowired
    private ServiceMapper serviceMapper;
    @Autowired
    private UserMapper userMapper;
    @Autowired
    private Storage storage;
    @Autowired
    private TelegramService telegramService;

    @Override
    public void handle(Message message) {
        var id = message.getFrom().getId();

        if (storage.contain(id)){
            return;
        }

        if (userMapper.getById(id) != null  && !telegramService.isAdmin(id)){
            getTime().apply(id)
                    .forEach(it -> send(it, message.getFrom()));
            return;
        }

        if (!telegramService.isAdmin(id)){
            return;
        }

        serviceMapper.getIds(userMapper.getById(id).getOrganizationId())
                .forEach(it -> getTime().apply(it)
                        .forEach(report -> send(report, message.getFrom())));
    }

    @Override
    public Role getRequiredRole() {
        return Role.READ;
    }

    protected abstract Function<Long, List<File>> getTime();


    private void send(File file, User from){
        SendDocument sendDocument = SendDocument
                .builder()
                .chatId(from.getId())
                .document(new InputFile(file))
                .build();
        sender.send(sendDocument, from);
        //noinspection ResultOfMethodCallIgnored
        file.delete();
    }

}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/telegram/commandHandler/statisticCommand/BaseReport.kt
package me.centralhardware.znatoki.telegram.statistic.telegram.commandHandler.statisticCommand

import me.centralhardware.znatoki.telegram.statistic.entity.Role
import me.centralhardware.znatoki.telegram.statistic.mapper.ServiceMapper
import me.centralhardware.znatoki.telegram.statistic.mapper.UserMapper
import me.centralhardware.znatoki.telegram.statistic.service.TelegramService
import me.centralhardware.znatoki.telegram.statistic.telegram.TelegramSender
import me.centralhardware.znatoki.telegram.statistic.telegram.commandHandler.CommandHandler
import me.centralhardware.znatoki.telegram.statistic.telegram.fsm.Storage
import me.centralhardware.znatoki.telegram.statistic.userId
import org.springframework.stereotype.Component
import org.telegram.telegrambots.meta.api.methods.send.SendDocument
import org.telegram.telegrambots.meta.api.objects.InputFile
import org.telegram.telegrambots.meta.api.objects.Update
import java.io.File

@Component
abstract class BaseReport(
    private val serviceMapper: ServiceMapper,
    private val storage: Storage,
    private val telegramService: TelegramService,
    sender: TelegramSender,
    userMapper: UserMapper
) : CommandHandler(sender, userMapper) {

    override fun handle(update: Update) {
        val id = update.userId()

        if (storage.contain(id)) { return }

        if (userMapper.getById(id) != null && !telegramService.isAdmin(id)) {
            getTime().invoke(id).forEach{
                send(it, update.userId())
            }
            return
        } else if (telegramService.isAdmin(id)){
            serviceMapper.getIds(userMapper.getById(id)!!.organizationId).forEach {
                getTime().invoke(it).forEach { send(it, update.userId()) }
            }
        }

    }

    override fun getRequiredRole(): Role = Role.READ

    protected abstract fun getTime(): (Long) -> List<File>


    private fun send(file: File, chatId: Long) {
        val sendDocument = SendDocument.builder()
            .chatId(chatId)
            .document(InputFile(file))
            .build()
        sender.send{ execute(sendDocument) }
        file.delete()
    }

}